### DISCUSSION TITLE
Runtime error "apply transform"

### DISCUSSION THREAD

**Discussion by vikashg** [Project-MONAI/MONAI#4121] [Q&A]
Hello everyone, 
I wrote a custom transform 
```
class ExtractValues(Transform):
	def __init__(self, key1: str) -> None:
		super().__init__(self)
		self.key = key1

	def __call__(self, data: List[Mapping[Hashable, NdarrayOrTensor]], *args, **kwargs):
		result = []
		for _data in data:
			print(_data)
			out = _data[self.key]
			result.append(out)
		return result
```
What this transform does is basically extract some values from a dictionary based on the key provided. When I compose it with another custom transform which reads a json file. I compose it using a transform chain 
```
           annotation_transform_list = Compose([LoadJson(labels]), ExtractValues(key1='points')])
```

I get the following error. 
```
   for xm, ym in train_loader:
  File "/opt/conda/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 521, in __next__
    data = self._next_data()
  File "/opt/conda/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 1203, in _next_data
    return self._process_data(data)
  File "/opt/conda/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 1229, in _process_data
    data.reraise()
  File "/opt/conda/lib/python3.8/site-packages/torch/_utils.py", line 438, in reraise
    raise exception
RuntimeError: Caught RuntimeError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/workspace/MONAI/monai/transforms/transform.py", line 88, in apply_transform
    return [_apply_transform(transform, item, unpack_items) for item in data]
  File "/workspace/MONAI/monai/transforms/transform.py", line 88, in <listcomp>
    return [_apply_transform(transform, item, unpack_items) for item in data]
  File "/workspace/MONAI/monai/transforms/transform.py", line 53, in _apply_transform
    return transform(parameters)
  File "Test.py", line 99, in __call__
    out = _data['points']
TypeError: string indices must be integers

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/torch/utils/data/_utils/worker.py", line 287, in _worker_loop
    data = fetcher.fetch(index)
  File "/opt/conda/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 49, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/opt/conda/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 49, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "GeometricTransfroms.py", line 45, in __getitem__
    return self.image_transform_list(self.imagelist[index]), self.annotation_transform_list(self.annotationlist[index])
  File "/workspace/MONAI/monai/transforms/compose.py", line 173, in __call__
    input_ = apply_transform(_transform, input_, self.map_items, self.unpack_items, self.log_stats)
  File "/workspace/MONAI/monai/transforms/transform.py", line 113, in apply_transform
    raise RuntimeError(f"applying transform {transform}") from e
RuntimeError: applying transform <__main__.ExtractValues object at 0x7f2a3c3889a0>
```

I am not sure what I am doing wrong. My other custom transforms never give such errors. Any thoughts would be really helpful. I will try to share the relevant code if needed. but I think the RunTimeError should giveaway the problems for experts @ericspod @rijobro 
Thanks 
Vikash

**Comment by Nic-Ma** [comment]
Hi @vikashg ,

I think that because your transform output is not a dictionary anymore, so they can't be composed.
Please ensure the transform output is a dictionary or a list of dictionaries and try again?

Thanks.

**Comment by vikashg** [comment]
Hi @Nic-Ma, 
Got it. I think any transform outputs are actually dictionary. Even for LoadImage the output is actually dictionary we just choose the output to be just image by setting the flag image_only=True. Am I correct in thinking so ?

Thanks sooo much this saved me soo much time.

**Reply by Nic-Ma** [reply]
Yes, it must be dictionary data.

Thanks.

**Reply by vikashg** [reply]
Hi Nick, 
So following up on your comment. I tried to look into the outputs of `LoadImage()` closely. If I do this
```
a = LoadImage()(list_of_files)
```
The output `a` is a list of 2 elements, the first one contains an array of shape [num_images, dimX, dimY, ...] However, the secondone contains a list of image attributes mostly pertaining to the first element in the `list_of_files`. I can say that because it prints the `filename_or_obj`. In an ideal case shouldn't the second element of a which contains the attributes should contain a list of attributes pertaining to all the images in `list_of_files`. Am I missing something here ?
Thanks 
Vikash

**Reply by Nic-Ma** [reply]
Hi @vikashg ,

When loading a list of files, they are stacked together at a new dimension as the first dimension, and the meta data of the first image is used to present the output meta data. You can find more details here:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/data/image_reader.py#L428

Thanks.
